src/v/cluster/drain_manager.h

```diff
@@ -102,6 +105,9 @@ class drain_manager : public ss::peering_sharded_service<drain_manager> {
     ssx::semaphore _sem{0, "c/drain-mgr"};
     drain_status _status;
     ss::abort_source _abort;
+    std::queue<operation_type> _requests;
```

Comment:
would simple mutex be sufficient to serialize calls to draining/restore ? It seems more complicated now that it was

Reply:
Code-wise yes, there is more code now but if you track the state changes, fibers changing it and their interaction then the coupling between active fibers got reduced, we pack less different meaning into the same state and previously hidden relationships became explicit overall reducing complexity.

For example:
- Have you noticed that `_draining` used to mean both a request, an ongoing process and a final state?
- Can you explain why we had `_sem.consume(_sem.available_units());` while `wait()` is supposed to do it automatically?

Now:
- `_draining` is true only when the maintenence mode is on and draining is ongoing
- `_drained` is true only when the maintenence mode is on and draining has finished
- `_requests` contains a list of requests for the task-fiber
- All state but request related is modified only by the task-fiber

Decoupling helps to reason about the system by focusing on one of its components without worrying about the side effects.

PR doesn't fundamentally change the current architecture of drain_manager:
- `task` fiber is responsible for taking actions
- `drain()` / `restore()` only schedule an action and return control flow before it is executed
- later invocation of `drain()` / `restore()` overwrite previously scheduled tasks

It's possible to achieve similar behavior with mutex but I didn't want to rewrite the drain_manager completely. It's a common pattern (producer / consumer?) to use channels for communication between threads / fiber. Here `_requests` acts as a task queue and semaphore's `signal()` / `wait()` for signalling similar to java's `notify` / `wait`.

Reply:
Here the semaphore is used as a condition variable, `wait()` only consumes a single unit while `_sem.consume(_sem.available_units());` consume rest of them hence only single task can be executed. I would say simplest way to achieve the desired semantics is to make `drain` and `restore` methods synchronous, they do not have to return future. This way we will have last request wins semantics not introducing a queue

Reply:
i think i undestand it now. @dotnwat can you take a quick look as it is your code

Reply:
synced with @mmaslankaprv and agreed that `_draining` has overloaded meaning which should be fixed; also found a common ground that the queue is an overkill and there is a chance of fixing it with an extra variable or two